### PR TITLE
fix: training_step_end doesn't work as stated in docs

### LINKED ISSUE #8105: training_step_end doesn't work as stated in docs
## 🐛 Bug

I'm training a model on multiple GPUs using the `dp` accelerator. I want to combine the outputs using the function `training_step_end`. However, unlike stated in the docs, in that function I don't have access to the outputs from all GPUs but just one.

## Please reproduce using the BoringModel

https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing

### To Reproduce

I can't 100% reproduce the error with BoringModel because I don't have access to (multiple) GPUs, so the behaviour of `dp` can't be tested. 
But when I run this on a 4 GPU machine, the print statement in `training_step_end` reveal that it just gets a dictionary with a logit/labels tensor of length 10, instead of 4 logit/labels tensors of length 10:

```
Epoch 10:  74%|███████▍  | 7428/10000 [01:06<00:22, 112.52it/s, loss=0.7, v_num=0]  {'logits': tensor([[-0.5810, -0.8195],
2021-06-23T15:50:47.151068998Z:         [-0.8803, -0.5356],
2021-06-23T15:50:47.151072929Z:         [-0.4559, -1.0047],
2021-06-23T15:50:47.151076585Z:         [-0.7245, -0.6628],
2021-06-23T15:50:47.151080187Z:         [-0.9753, -0.4733],
2021-06-23T15:50:47.151083827Z:         [-0.6311, -0.7593],
2021-06-23T15:50:47.151088058Z:         [-0.5547, -0.8539],
2021-06-23T15:50:47.151092906Z:         [-0.7836, -0.6102],
2021-06-23T15:50:47.151097823Z:         [-0.7722, -0.6199],
2021-06-23T15:50:47.151104739Z:         [-0.7415, -0.6470]], device='cuda:0', grad_fn=<GatherBackward>), 'labels': tensor([0, 1, 0, 1, 1, 1, 1, 0, 0, 0], device='cuda:0')}
``` 
Use following [**BoringModel**](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing) 

### Expected behavior

According to the [docs](https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html), and in particular the code snippet below coming from the docs, I would expect `batch_parts` to be a tuple containing `n` elements where `n` is the number of GPUs used. However, as can be seen in BoringModel, and what I'm getting when training on a machine with 4 GPUs, is that `batch_parts` is just a dictionary containing the output of **one** `training_step` (and not 4 training steps)

```python
def training_step(self, batch, batch_idx):
    x, y = batch
    y_hat = self.model(x)
    loss = F.cross_entropy(y_hat, y)
    pred = ...
    return {'loss': loss, 'pred': pred}

def training_step_end(self, batch_parts):
    gpu_0_prediction = batch_parts[0]['pred']
    gpu_1_prediction = batch_parts[1]['pred']

    # do something with both outputs
    return (batch_parts[0]['loss'] + batch_parts[1]['loss']) / 2

def training_epoch_end(self, training_step_outputs):
   for out in training_step_outputs:
       # do something with preds
```

### Environment

```
* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           10.2
* Packages:
	- numpy:             1.19.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.9.0+cu102
	- pytorch-lightning: 1.3.7post0
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.7.10
	- version:           #1 SMP Sat Jun 5 09:50:34 PDT 2021
```

### PR DESCRIPTION
**Pull request by satishjasthi**
This PR fixes the bugs in the documentation wrt issue 8105. This is a documentation based PR

Fixes #8105 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
     - I couldn't because constraints on my system ram, I have only 4GB ram and when I run `make test` it freeze my system because of full utilisation of available memory
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Yes 😀

### CODE REVIEW

File: docs/source/common/lightning_module.rst [hunk 1]
```diff
@@ -279,11 +279,16 @@ In this case, implement the `training_step_end` method
          return {'loss': loss, 'pred': pred}
 
      def training_step_end(self, batch_parts):
-         gpu_0_prediction = batch_parts[0]['pred']
-         gpu_1_prediction = batch_parts[1]['pred']
+         # predictions from each GPU
+         predictions = batch_parts['pred]
+         # losses from each GPU
+         losses = batch_parts['loss]
```
Comment (awaelchli) [inline docs/source/common/lightning_module.rst:285]:
```suggestion
         predictions = batch_parts['pred']
         # losses from each GPU
         losses = batch_parts['loss']
```

File: docs/source/common/lightning_module.rst [hunk 2]
```diff
@@ -383,11 +388,16 @@ In this case, implement the `validation_step_end` method
          return {'loss': loss, 'pred': pred}
 
      def validation_step_end(self, batch_parts):
-         gpu_0_prediction = batch_parts.pred[0]['pred']
-         gpu_1_prediction = batch_parts.pred[1]['pred']
+         # predictions from each GPU
+         predictions = batch_parts['pred]
+         # losses from each GPU
+         losses = batch_parts['loss]
```
Comment (awaelchli) [inline docs/source/common/lightning_module.rst:394]:
```suggestion
         predictions = batch_parts['pred']
         # losses from each GPU
         losses = batch_parts['loss']
```

**Review by awaelchli — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: docs/source/common/lightning_module.rst (hunk 1 above)
Comment (satishjasthi) [reply docs/source/common/lightning_module.rst]: Thank you for pointing out, my bad 🤦‍♂️I had resolved this with a new commit

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]
